pkg/receiver/jobreceiver/command/command_test.go

```diff
@@ -0,0 +1,183 @@
+		outC := eventualOutput(echo)
+		resp, err := echo.Run()
+		require.NoError(t, err)
+		assert.Equal(t, 0, resp.Status)
+		require.NoError(t, err)
+		assert.Contains(t, <-outC, "hello world")
+	})
+
+	// Command exits non-zero
+	t.Run("exit code 1", func(t *testing.T) {
+		exitCmd := NewExecution(ctx, withTestHelper(t, ExecutionRequest{Command: "exit", Arguments: []string{"1"}}))
+		resp, err := exitCmd.Run()
+		require.NoError(t, err)
+		assert.Equal(t, 1, resp.Status)
+		exitCmd = NewExecution(ctx, withTestHelper(t, ExecutionRequest{Command: "exit", Arguments: []string{"33"}}))
+		resp, err = exitCmd.Run()
+		require.NoError(t, err)
+		assert.Equal(t, 33, resp.Status)
+	})
+
+	// Command exceeds timeout
+	t.Run("exceeds timeout", func(t *testing.T) {
+		timeout := time.Millisecond * 100
+		sleepCmd := NewExecution(ctx, withTestHelper(t, ExecutionRequest{Command: "sleep", Arguments: []string{"1m"}, Timeout: timeout}))
+		done := make(chan struct{})
+		go func() {
+			resp, err := sleepCmd.Run()
+			assert.NoError(t, err)
+			assert.NotEqual(t, OKExitStatus, resp.Status)
+			assert.LessOrEqual(t, timeout, resp.Duration)
+			close(done)
+		}()
+		select {
+		case <-time.After(5 * time.Second):
+			t.Errorf("command timeout exceeded but was not killed")
+		case <-done:
+			// okay
+		}
+	})
+	// Command exceeds timeout with child process
+	t.Run("exceeds timeout with child", func(t *testing.T) {
+		timeout := time.Millisecond * 100
+		sleepCmd := NewExecution(ctx, withTestHelper(t, ExecutionRequest{Command: "fork", Arguments: []string{"sleep", "1m"}, Timeout: timeout}))
+		done := make(chan struct{})
+		go func() {
+			resp, err := sleepCmd.Run()
+			assert.NoError(t, err)
+			assert.NotEqual(t, OKExitStatus, resp.Status)
+			assert.LessOrEqual(t, timeout, resp.Duration)
+			close(done)
+		}()
+		select {
+		case <-time.After(5 * time.Second):
+			t.Fatal("command timeout exceeded but was not killed")
+		case <-done:
+			// okay
+		}
+	})
+
+	// Invocation cannot be spuriously re-invoked
+	t.Run("cannot be ran twice", func(t *testing.T) {
+		echo := NewExecution(ctx, withTestHelper(t, ExecutionRequest{Command: "echo", Arguments: []string{"hello", "world"}}))
+		outC := eventualOutput(echo)
+		echo.Run()
+		_, err := echo.Run()
+		assert.Error(t, err)
+		assert.Contains(t, <-outC, "hello world")
+	})
+}
+
+// withTestHelper takes an ExecutionRequest and adjusts it to run with the
+// test binary. TestMain will handle emulating the command.
+func withTestHelper(t *testing.T, request ExecutionRequest) ExecutionRequest {
+	t.Helper()
+
+	request.Arguments = append([]string{request.Command}, request.Arguments...)
+	request.Command = os.Args[0]
+	return request
+}
+
+//nolint:all
+func eventualOutput(i *Execution) <-chan string {
+	out := make(chan string, 1)
+	stdout, _ := i.Stdout()
+	stderr, _ := i.Stderr()
+	go func() {
+		var buf syncBuffer
+		io.Copy(&buf, stdout)
+		io.Copy(&buf, stderr)
```

Comment:
Why do we need synchronization here?

Reply:
Good catch! We don't - just an artifact of one revision of this.